addons/account/models/account_move.py

```diff
@@ -1793,6 +1794,75 @@ def _build_credit_warning_message(self, record, updated_credit):
         msg += ': %s' % formatLang(self.env, updated_credit, currency_obj=record.company_id.currency_id)
         return msg
 
+    @api.depends('ref', 'move_type', 'partner_id', 'invoice_date')
+    def _compute_duplicated_ref_ids(self):
+        # The computed field only appears in the move view form and thus is supposed
+        # to only be computed for one move at a time which makes the search in loop "ok"
+        move_to_duplicate_move = self._fetch_duplicate_supplier_reference()
+        for move in self:
+            move.duplicated_ref_ids = move_to_duplicate_move.get(move, self.env['account.move'])
+
+
+    def _fetch_duplicate_supplier_reference(self, only_posted=False):
+        move_to_duplicate_moves = defaultdict(lambda x: self.env['account.move'])
+        moves = self.filtered(lambda m: m.is_purchase_document() and m.ref)
+        if not moves:
+            return move_to_duplicate_moves
+
+        # this case deals with a move created in UI having id being a NEW_ID object
+        if len(moves) == 1 and not moves.id:
+            duplicates = self.env["account.move"].search(
+                [
+                    ('company_id', '=', moves.company_id.id),
+                    ('move_type', '=', moves.move_type),
+                    ('commercial_partner_id', '=', moves.commercial_partner_id.id),
+                    ('ref', '=', moves.ref)
+                ]
+                + [
+                    ('invoice_date', '=', moves.invoice_date),
+                    ('state', '=', 'posted'),
+                ] if only_posted else []
+            )
+            move_to_duplicate_moves.update({
+                moves: duplicates
+            })
+            return move_to_duplicate_moves
+
+        self.env["account.move"].flush([
+            "company_id",
+            "partner_id",
+            "commercial_partner_id",
+            "ref",
+            "move_type",
+            "invoice_date",
+            "state",
+        ])
+        self.env.cr.execute("""
+            SELECT
+                   move.id,
+                   array_agg(duplicate_move.id)
+              FROM account_move move
+              JOIN account_move AS duplicate_move ON
+                   move.company_id = duplicate_move.company_id
+               AND move.commercial_partner_id = duplicate_move.commercial_partner_id
+               AND move.ref = duplicate_move.ref
+               AND move.move_type = duplicate_move.move_type
+               AND move.id != duplicate_move.id
+               AND (move.invoice_date = duplicate_move.invoice_date OR NOT %(only_posted)s)
+               AND (duplicate_move.state = 'posted' OR NOT %(only_posted)s)
+             WHERE move.id IN (%(moves)s)
+             GROUP BY move.id
+        """, {
+            "only_posted": only_posted,
+            "moves": tuple(moves.ids),
+        })
```

Comment:
Maybe something like this?
I mostly wanted to try it, not sure if it's better
```suggestion
        dup_query = expression.expression([('id', 'in', self.ids)], self).query
        duplicate_alias = dup_query.join(
            lhs_alias='account_move', lhs_column='ref',
            rhs_table='account_move', rhs_column='ref',
            link='duplicate',
            extra="""
                {lhs}.commercial_partner_id = {rhs}.commercial_partner_id
                AND {lhs}.company_id = {rhs}.company_id
                AND {lhs}.move_type = {rhs}.move_type
            """
        )
        if only_posted:
            dup_query.add_where(f"{duplicate_alias}.state = 'posted'")
            dup_query.add_where(f"{duplicate_alias}.invoice_date = account_move.invoice_date")
        query_str, params = dup_query.select('account_move.id', f'array_agg({duplicate_alias}.id)')
        query_str += " GROUP BY account_move.id"

        self.env.cr.execute(query_str, params)